language/move-symbol-pool/src/lib.rs

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! A global, uniqued cache of strings that is never purged.
+//!
+//! This module provides storage for strings that are meant to remain in use for
+//! the entire running duration of a program. Strings that are stored in this
+//! global, static cache are never evicted, and so the memory consumed by them
+//! can only ever grow.
+//!
+//! The strings can be accessed via the [`Symbol`] type, which acts like a
+//! pointer to the underlying cached string value.
+//!
+//! NOTE: If you're looking for a `#[forbid(unsafe_code)]` attribute here, you
+//! won't find one: symbol-pool (and its inspiration, servo/string-cache) uses
+//! `unsafe` Rust in order to store and dereference `Symbol` pointers to
+//! strings.
+//!
+//! [`Symbol`]: crate::move_symbol_pool::Symbol
+
+use once_cell::sync::Lazy;
+use serde::{Deserialize, Serialize};
+use std::{cmp::Ordering, collections::HashSet, fmt, num::NonZeroU64, sync::Mutex};
+
+type SymbolData = Box<str>;
+static SYMBOL_POOL: Lazy<Mutex<HashSet<SymbolData>>> = Lazy::new(|| Mutex::new(HashSet::new()));
+
+/// Represents a string that has been cached.
+///
+/// A `Symbol` represents a string; it is not the string data itself. As such
+/// this representation can implement `Copy` and other traits that some string
+/// types cannot.
+///
+/// The strings that `Symbol` types represent are added to the global cache as
+/// the `Symbol` are created.
+///
+/// ```
+///# use crate::move_symbol_pool::Symbol;
+/// let s1 = Symbol::from("hi"); // "hi" is stored in the global cache
+/// let s2 = Symbol::from("hi"); // "hi" is already stored, cache does not grow
+/// assert_eq!(s1, s2);
+/// ```
+///
+/// Use the method [`as_str()`] to access the string value that a `Symbol`
+/// represents. `Symbol` also implements the [`Display`] trait, so it can be
+/// printed as an ordinary string would. This makes it easier to use with
+/// crates that print strings to a terminal, such as codespan.
+///
+/// ```
+///# use crate::move_symbol_pool::Symbol;
+/// let message = format!("{} {}",
+///     Symbol::from("hello").as_str(),
+///     Symbol::from("world"));
+/// assert_eq!(message, "hello world");
+/// ```
+///
+/// [`as_str()`]: crate::move_symbol_pool::Symbol::as_str
+/// [`Display`]: std::fmt::Display
+#[derive(Clone, Copy, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
+pub struct Symbol(NonZeroU64);
+
+impl Symbol {
+    pub fn as_str(&self) -> &str {
+        let ptr = self.0.get() as *const SymbolData;
+        unsafe { &(*ptr) }
+    }
+}
+
+impl From<&str> for Symbol {
```

Comment:
I kind of like this a lot more than using all the Rust traityness used by string_cache. All of those traits tend to hide what is going on

Having said that, it might be useful to do the optimization for `String` v. `&str` for `From` as you don't need to call `to_owned` for the `String` case